### PR TITLE
Some parser fixes

### PR DESCRIPTION
**Pull request by FaresAtef1**
This PR consumes the semicolon (if it exists) for certain statements and expressions, includes new test cases, and fix a typo.

### Modified Statements:
- Return
- Break
- Continue

### Modified Expressions:
- Delete
- New
- Unary Postfix

### Testing
- Add new switch case tests.
- Complete for loop tests.
- Utilize Google Mock for `ASSERT_THAT` with new matchers: `IsDeclaration`, `IsCast`, `IsBinaryOperation`, `IsUnaryPostfixOperator`, `IsUnaryPrefixOperator` and `IsStatementBlock`.
- Add new for loop tests.

### New Functions
- `NodeToString` in `AST` class to get the node type as a string.
- `StorageToString` in `AST::DeclarationStatement` class to get the storage type as a string.
- `KindToString` in `AST::CastExpression` class to get the cast type as a string.

> [!note]
> I used clang-format for the `parser-test.cpp` file.

### CODE REVIEW

File: CompilerSource/parsing/ast.cpp [hunk 1]
```diff
@@ -157,4 +157,52 @@ void AST::DeclarationStatement::RecursiveSubVisit(Visitor &visitor) {
   
 }
 
+const std::vector<std::string> AST::NodesNames = [](){
+  std::vector<std::string> res;
+  int nsize = static_cast<int>(NodeType::INITIALIZER)+1;
+  res.resize(nsize);
+
+  #define REGISTER(name) [[fallthrough]]; case name: res[(int)AST::name] = \
+    std::string(#name).substr(std::string(#name).rfind(':') + 1)
```
Comment (FaresAtef1) [inline CompilerSource/parsing/ast.cpp:166]: To get rid of the starting `AST::`.

File: CommandLine/emake-tests/Parsing/parser-test.cpp [hunk 2]
```diff
@@ -582,55 +659,55 @@ TEST(ParserTest, TemporaryInitialization_1) {
...
-  ASSERT_NE(decl1->init, nullptr);
-  ASSERT_EQ(decl1->init->type, AST::NodeType::INITIALIZER);
-  ASSERT_EQ(decl1->init->is_variadic, false);
-  ASSERT_EQ(decl1->init->kind, AST::Initializer::Kind::ASSIGN_EXPR);
-  auto *init = std::get<AST::AssignmentInitNode>(decl1->init->initializer).get();
-  ASSERT_EQ(init->kind, AST::AssignmentInitializer::Kind::EXPR);
-  auto *expr = std::get<std::unique_ptr<AST::Node>>(init->initializer).get();
-  assert_identifier_is(expr, "nullptr");
-}
+// TEST(ParserTest, TemporaryInitialization_2) {
```
Comment (JoshDreamland) [inline CommandLine/emake-tests/Parsing/parser-test.cpp:668]: Another good place for C-style comments, if you would.

**Review by JoshDreamland — COMMENTED** [review]
This is looking good, thank you! I'll help you fix the parsing bug after we merge this. I'd like to revert most of those comment lines (by switching to C-style comments) first, though.

File: CommandLine/emake-tests/Parsing/parser-test.cpp [hunk 3]
```diff
@@ -125,9 +125,7 @@ TEST(ParserTest, SizeofType) {
   ASSERT_TRUE(std::holds_alternative<FullType>(sizeof_->argument));
 
   auto &value = std::get<FullType>(sizeof_->argument);
-  auto has_value = [&value](jdi::typeflag *builtin) -> bool {
-    return (value.flags & builtin->mask) == builtin->value;
-  };
+  auto has_value = [&value](jdi::typeflag *builtin) -> bool { return (value.flags & builtin->mask) == builtin->value; };
```
[the remark this answers is not in the document]
Comment (FaresAtef1) [reply CommandLine/emake-tests/Parsing/parser-test.cpp:128]: Okay, I will return them.

File: CompilerSource/parsing/parser.cpp [hunk 4]
```diff
@@ -1475,6 +1475,10 @@ std::unique_ptr<AST::Node> TryParseNewExpression(bool is_global) {
   is_array = !type.decl.components.empty() &&
              type.decl.components.begin()->kind == DeclaratorNode::Kind::ARRAY_BOUND;
 
+  if(token.type == TT_SEMICOLON) {
```
[the remark this answers is not in the document]
Comment (FaresAtef1) [reply CompilerSource/parsing/parser.cpp:1478]: I've already done that, but I didn't create separate tests. I will create them and push the changes. It's quite fast, thanks to your `ASSERT_THAT` idea.

**Review by JoshDreamland — APPROVED** [review]
Looking great, thank you!

**Pull request merged**